server/src/main/java/org/opensearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -618,9 +621,20 @@ protected void onShardResult(Result result, SearchShardIterator shardIt) {
         if (logger.isTraceEnabled()) {
             logger.trace("got first-phase result from {}", result != null ? result.getSearchShardTarget() : null);
         }
+        this.setPhaseResourceUsages();
         results.consumeResult(result, () -> onShardResultConsumed(result, shardIt));
     }
 
+    public void setPhaseResourceUsages() {
+        ThreadContext threadContext = searchRequestContext.getThreadContextSupplier().get();
+        if (threadContext != null) {
+            List<String> taskResourceUsages = threadContext.getResponseHeaders().get(TASK_RESOURCE_USAGE);
+            if (taskResourceUsages != null && taskResourceUsages.size() > 0) {
+                searchRequestContext.recordPhaseResourceUsage(taskResourceUsages.get(0));
+            }
```

Comment:
Because we only inject 1 element to this response header, and response headers is a `Map<String, List<String>>`. We only care about the first one. Do you think we should log an debug/error log if the size is greater than 1? Since that will be an unexpected scenario.

Reply:
If that is hard assumption, we can always assert that

Reply:
IMO, updating to assert might have a bigger impact on the process. With the current logic we are more tolerant on the potentiall errors.